File: src/mc_server_manager/config/dotenv_loader.py
from __future__ import annotations

from pathlib import Path
from typing import Iterable

from mc_server_manager.config.settings import SftpSettings


class DotEnvLoader:
    required_keys = (
        "SFTP_HOST",
        "SFTP_PORT",
        "SFTP_USERNAME",
        "SFTP_PASSWORD",
        "SFTP_SERVER_ROOT",
    )

    def default_paths(self) -> list[Path]:
        candidates = [Path.cwd() / ".env", Path(__file__).resolve().parents[3] / ".env"]
        unique_paths: list[Path] = []
        seen: set[Path] = set()
        for candidate in candidates:
            resolved = candidate.resolve()
            if resolved not in seen:
                unique_paths.append(resolved)
                seen.add(resolved)
        return unique_paths

    def load(self, candidate_paths: Iterable[Path] | None = None) -> SftpSettings:
        candidate_paths = list(candidate_paths or self.default_paths())
        env_path = next((path for path in candidate_paths if path.is_file()), None)
        if env_path is None:
            checked = ", ".join(str(path) for path in candidate_paths)
            raise FileNotFoundError(f"Expected a .env file. Checked: {checked}")

        values: dict[str, str] = {}
        for raw_line in env_path.read_text(encoding="utf-8").splitlines():
            line = raw_line.strip()
            if not line or line.startswith("#"):
                continue

            if "=" not in line:
                continue

            key, value = line.split("=", 1)
            values[key.strip()] = value.strip().strip('"')

        missing = [key for key in self.required_keys if not values.get(key)]
        if missing:
            raise ValueError(f"Missing required .env keys: {', '.join(missing)}")

        try:
            port = int(values["SFTP_PORT"])
        except ValueError as exc:
            raise ValueError("SFTP_PORT must be a valid integer.") from exc

        return SftpSettings(
            host=values["SFTP_HOST"],
            port=port,
            username=values["SFTP_USERNAME"],
            password=values["SFTP_PASSWORD"],
            server_root=values["SFTP_SERVER_ROOT"],
        )

File: src/McServerManager.Infrastructure/Environment/DotEnvLoader.cs
using McServerManager.Application.Configuration;

namespace McServerManager.Infrastructure.Environment;

public sealed class DotEnvLoader : IEnvironmentLoader
{
    public SftpSettings Load(string path)
    {
        if (!File.Exists(path))
        {
            throw new FileNotFoundException($"Expected a .env file at '{path}'. Copy .env.example and fill in your SFTP settings.");
        }

        var values = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        foreach (var rawLine in File.ReadAllLines(path))
        {
            var line = rawLine.Trim();
            if (string.IsNullOrWhiteSpace(line) || line.StartsWith("#", StringComparison.Ordinal))
            {
                continue;
            }

            var separatorIndex = line.IndexOf('=');
            if (separatorIndex <= 0)
            {
                continue;
            }

            var key = line[..separatorIndex].Trim();
            var value = line[(separatorIndex + 1)..].Trim().Trim('"');
            values[key] = value;
        }

        var missingKeys = RequiredKeys.Where(key => !values.TryGetValue(key, out var value) || string.IsNullOrWhiteSpace(value)).ToArray();
        if (missingKeys.Length > 0)
        {
            throw new InvalidOperationException($"The .env file is missing required keys: {string.Join(", ", missingKeys)}");
        }

        if (!int.TryParse(values["SFTP_PORT"], out var port))
        {
            throw new InvalidOperationException("SFTP_PORT must be a valid integer.");
        }

        return new SftpSettings(
            values["SFTP_HOST"],
            port,
            values["SFTP_USERNAME"],
            values["SFTP_PASSWORD"],
            values["SFTP_SERVER_ROOT"]);
    }

    private static readonly string[] RequiredKeys =
    [
        "SFTP_HOST",
        "SFTP_PORT",
        "SFTP_USERNAME",
        "SFTP_PASSWORD",
        "SFTP_SERVER_ROOT",
    ];
}
